File: runtime/java/units/wasi/fd_datasync.java
// force(false) flushes the file content without forcing metadata, the
// datasync guarantee.
int wasi_fd_datasync(int fd) {
    Object e = fds.get(fd);
    if (!(e instanceof Handle)) {
        return WASI_BADF;
    }
    try {
        ((Handle) e).ch.force(false);
    } catch (java.io.IOException ex) {
        return WASI_IO;
    }
    return WASI_OK;
}

File: runtime/go/units/wasi/fd_datasync.go
// Go's os package exposes no fdatasync (and it is absent on macOS anyway), so
// fall back to a full fsync, a superset of the datasync guarantee (mirroring
// the Python backend's fallback).
func (w *WASI) wasi_fd_datasync(fd uint32) uint32 {
    f, ok := w.fds[fd].(*os.File)
    if !ok {
        return wasiBadf
    }
    if err := f.Sync(); err != nil {
        return wasiIo
    }
    return wasiOk
}
